Add random tick events

ChibChi opened the pull request:
This pull request adds random tick events for blocks. Tick events can be added by using the `tickEvent` trait and contain the `name` of the callback function and the probability ( `chance`).

Resolves: https://github.com/PixelGuys/Cubyz/issues/77

ChibChi commented on src/server/world.zig:
```diff
@@ -927,6 +933,39 @@ pub const ServerWorld = struct { // MARK: ServerWorld
+		if(_chunk) |ch| {
+			var iter = ch.super.blockPosToTickableBlockMap.iterator();
+
+			while(iter.next()) |entry| {
```
Ah, my bad. I thought this is more like hoppers/redstone, which will be a different thing i guess.

ChibChi commented on src/blocks.zig:
```diff
@@ -369,32 +387,83 @@ pub const Block = packed struct { // MARK: Block
+	}
+
+	pub fn shouldTick(self: *const TickEvent, tickCount: i64) bool {
+		return self.tryEvery <= 1 or @mod(tickCount, @as(i64, @intCast(self.tryEvery))) == 0;
```
Is `tryEvery` even usefull? Only a few blocks can try to tick each chunk. This would add another layer of randomness, because the random chance has to trigger and the tick count has to overlap the delay.

IntegratedQuantum replied:
I think it would be useful if you want to make certain interactions slower than the (already slow) random tick speed. However we can always readd in the future it if we actually have such a use case.

ChibChi replied:
Yea, I removed it for now.

ChibChi commented on src/chunk.zig:
```diff
@@ -257,6 +257,9 @@ pub const Chunk = struct { // MARK: Chunk
 	blockPosToEntityDataMapMutex: std.Thread.Mutex,
 
+	blockPosToTickableBlockMap: std.AutoHashMapUnmanaged(u32, u32),
+	blockPosToTickableBlockMutex: std.Thread.Mutex,
```
What data structure would you suggest?
The only other option im seeing is a list.

IntegratedQuantum replied:
You could also not store them separately at all for now, and just pick a random block from the entire chunk volume each time.

IntegratedQuantum replied:
That's how it's usually done I think.

IntegratedQuantum commented on src/blocks.zig:
```diff
@@ -409,29 +409,41 @@ pub const TickFunctions = struct {
+		var tickEvents = _allocator.alloc(TickEvent, zon.toSlice().len);
 
-		for(zon.toSlice(), 0..) |tickEventZon, i| {
+		var n: usize = 0;
```
Please just use a `List` instead of emulating it.
Then you can just use `toOwnedSlice` in the end.

ChibChi commented on src/blocks.zig:
```diff
@@ -369,32 +387,83 @@ pub const Block = packed struct { // MARK: Block
+};
+
+pub const TickEvent = struct {
+	function: ?*const TickFunction,
```
Done.
If the callback can't be found, the slice resizes to prevent empty spots.

ChibChi commented:
I changed the block selection to a random block within the chunk. I also added `tickDistance` and `tickSpeed`. `tickDistance` is the radius around the player where tick events can happen and `tickSpeed` is the amount of block checks times 3. They have pretty bad names, now that i think about it.

IntegratedQuantum commented:
> `tickDistance`

This already exists in the form of `settings.simulationDistance`, conveniently that's already used when determining which chunks to put in the hashmap in the first place, so you don't need to do anything here.

> `tickSpeed` is the amount of block checks times 3

That is a rather odd convention, it should be specified as blocks per chunk with 3 being the default instead.

ChibChi commented:
> That is a rather odd convention, it should be specified as blocks per chunk with 3 being the default instead.

I generate 3 location from a single random number, which too be fair might be premature optimisation. Also only 3 blocks each tick sound very small, when a chunk is 32 blocks wide. I counted for around 12.

IntegratedQuantum commented:
Oh, I was not questioning the number, merely the way that you base your scale around it.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -428,6 +434,11 @@ pub const ServerWorld = struct { // MARK: ServerWorld
 	lastUnimportantDataSent: i64,
 	doGameTimeCycle: bool = true,
 
+	tickCount: i64 = 0,
```
What's tickCount for?

ChibChi replied:
I used to use it for `tryEvery`. I just forgot to remove it.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -948,6 +990,10 @@ pub const ServerWorld = struct { // MARK: ServerWorld
 				main.network.Protocols.genericUpdate.sendTimeAndBiome(user.conn, self);
 			}
 		}
+		if(self.lastTickTime + 50 < newTime) { // Tick very 50ms
```
I think this should be done every server tick, otherwise we'll end up with an uneven load between server ticks. You can adjust the tick rate accordingly.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -927,6 +939,36 @@ pub const ServerWorld = struct { // MARK: ServerWorld
+	fn tickBlocksInChunk(self: *ServerWorld, _chunk: ?*chunk.ServerChunk) void {
+		if(_chunk) |ch| {
+			for(0..self.tickSpeed) |_| {
+				const blockIndex: u32 = main.random.nextInt(u32, &main.seed);
```
If you used a `i32` directly here, you wouldn't need to cast below, and you could also save one mask.

ChibChi replied:
Wait, which mask can I save?

IntegratedQuantum replied:
Just ignore it, sorry, I forgot to remove that part.
Originally it said had `u15` in the comment, which would have saved one of the `&` below, but I left it out because llvm is kind of bad at generating code for partial integers.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -927,6 +939,36 @@ pub const ServerWorld = struct { // MARK: ServerWorld
+				const y: i32 = @intCast(blockIndex >> chunk.chunkShift & chunk.chunkMask);
+				const z: i32 = @intCast(blockIndex & chunk.chunkMask);
+
+				var block = ch.super.getBlock(x, y, z);
```
```suggestion
				const block = ch.super.getBlock(x, y, z);
```

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -927,6 +939,36 @@ pub const ServerWorld = struct { // MARK: ServerWorld
+	fn tick(self: *ServerWorld) void {
+		// tick blocks
+		var iter = ChunkManager.entityChunkHashMap.keyIterator();
+		while(iter.next()) |pos| {
```
This is not thread safe, another thread could modify the hashmap while you are iterating.

IntegratedQuantum commented on src/blocks.zig:
```diff
@@ -74,6 +77,7 @@ var _lodReplacement: [maxBlockCount]u16 = undefined;
 var _opaqueVariant: [maxBlockCount]u16 = undefined;
 var _friction: [maxBlockCount]f32 = undefined;
 var _allowOres: [maxBlockCount]bool = undefined;
+var _tickEvents: [maxBlockCount][]TickEvent = undefined;
```
I'm not convinced that we need multiple tick events. Could you give an example where we could need multiple tick events for one block?

ChibChi replied:
This was just my initial thought. This might be usefull for fire, to burn blocks in the area or run out of fuel. They might have different probabilities. To be fair thats the only example i can think of.

IntegratedQuantum replied:
For fire that should be a combined function anyways in my opinion.

ChibChi replied:
Yea, I made it an optional for now. This can be changed when its needed

IntegratedQuantum commented on src/blocks.zig:
```diff
@@ -372,32 +390,92 @@ pub const Block = packed struct { // MARK: Block
+
+	pub fn getFunctionPointer(id: []const u8) utils.CallbackError!*const TickFunction {
+		return super.getFunctionPointer(id);
+	}
```
Instead wrapping all the functions from `super`, why not just declare it outside the struct directly:
```
pub var tickFunctions: utils.NamedCallbacks(TickFunctions, TickFunction) = undefined;
```

ChibChi replied:
Done

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -928,6 +935,37 @@ pub const ServerWorld = struct { // MARK: ServerWorld
 	}
 
+	fn tickBlocksInChunk(self: *ServerWorld, _chunk: ?*chunk.ServerChunk) void {
+		if(_chunk) |ch| {
```
Please generally do early returns instead of long if's to avoid the extra indentation:
```suggestion
		const ch = _chunk orelse return;
```
Furthermore I think that this should probably be done at the call site without passing an optional to the function in the first place.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -928,6 +935,37 @@ pub const ServerWorld = struct { // MARK: ServerWorld
+				const y: i32 = blockIndex >> chunk.chunkShift & chunk.chunkMask;
+				const z: i32 = blockIndex & chunk.chunkMask;
+
+				const block = ch.super.getBlock(x, y, z);
```
Please use `ch.getBlock`.
Also this is not thread-safe, and `ch.getBlock` will report an error because of that.

ChibChi replied:
I hope it works now. Had some problems with dead locks and functions requiring the mutex to be unlocked, like the `ServerChunk.updateBlock()` gets or generates the chunk and wants to lock the entity chunk HashMap, but its already locked when I iterate through the entity chunks.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1884,3 +1884,70 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+
+pub fn NamedCallbacks(comptime Child: type, comptime Function: type) type {
+	return struct {
+		const Super = @This();
```
Why is this called `Super`?
General convention is to call the current struct `Self` if it cannot be referred to by name.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1884,3 +1884,70 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+		}
+
+		pub fn getFunctionPointer(self: *Super, id: []const u8) CallbackError!*const Function {
+			const pointer = self.hashMap.getPtr(id);
+			if(pointer == null) {
```
```suggestion
			const pointer = self.hashMap.getPtr(id) orelse {
```

IntegratedQuantum reviewed (changes requested):
Just some more smaller changes.

I'm also curious about the overall performance impact here.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -928,6 +935,38 @@ pub const ServerWorld = struct { // MARK: ServerWorld
+		while(iter.next()) |pos| {
+			if(ChunkManager.entityChunkHashMap.get(pos.*)) |entityChunk| {
+				const ch = entityChunk.getChunk() orelse continue;
+				entityChunk.increaseRefCount();
```
Alright, so some lessons on multi-threading:
- Doing refcounting while already holding the lock of the parent data structure is pointless. We do refcounting to allow us to keep using it on changes to the parent data structure.
- For refcounting you must increase the refcount while holding the lock to the containing data structure, afterwards you are free to decrease it outside the lock.

And here we don't want to hold the lock while executing the tick function, since as you discovered it causes a deadlock.

I think the correct solution would be to copy all chunks into a list first and increase all of their refcounts while holding the lock, and then afterwards you can actually go through them and execute the tick function without holding any lock.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -928,6 +935,38 @@ pub const ServerWorld = struct { // MARK: ServerWorld
+			const block = _chunk.getBlock(x, y, z);
+			_chunk.mutex.unlock();
+			if(block.tickEvent()) |event| {
+				event.tryRandomTick(block, _chunk, x, y, z);
```
I think you should still hold the lock of the chunk while entering the tick function, otherwise the block may change during the process of the tick function.
e.g. if the player breaks the block at the same time this might be used to duplicate the block, because it would now both drop the original block and apply the transformation.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1935,3 +1935,58 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+		pub fn getFunctionPointer(self: *Self, id: []const u8) CallbackError!*const Function {
+			const pointer = self.hashMap.getPtr(id) orelse {
+				if(id.len != 0)
+					return CallbackError.NotFound;
```
I think this is a bad design choice and this case should be handled differently at the call site, see my comment there.

IntegratedQuantum commented on src/blocks.zig:
```diff
@@ -123,7 +127,17 @@ pub fn register(_: []const u8, id: []const u8, zon: ZonElement) u16 {
-	_touchFunction[size] = TouchFunctions.getFunctionPointer(zon.get([]const u8, "touchFunction", ""));
+	_tickEvent[size] = TickEvent.loadFromZon(zon.getChild("tickEvent"));
+
+	const touchFunctionName = zon.get([]const u8, "touchFunction", "");
```
Instead of using `""` to signify that the zon entry did not exist, better use an optional.
This also gets rid of the weird and confusing error handling paths.
```suggestion
	const touchFunctionName = zon.get(?[]const u8, "touchFunction", null);
```

IntegratedQuantum commented on src/blocks.zig:
```diff
@@ -371,35 +388,59 @@ pub const Block = packed struct { // MARK: Block
+		const blockUpdates: [1]main.renderer.mesh_storage.BlockUpdate = .{newBlock};
+		for(userList) |user| {
+			main.network.Protocols.blockUpdate.send(user.conn, blockUpdates[0..]);
+		}
```
There is also another problem here:
This does not consider neighboring blocks.
E.g. if you have a torch that is connected to the current block and it gets replaced with air (e.g. leaf decay), then the torch wouldn't be removed.

A better solution would be to use `world.cmpxchgBlock` which handles this as well as the networking code.
You also wouldn't need to lock the chunk mutex anymore (→no chance for deadlocks in the future), since the function will compare the block first and only exchange it when it hasn't changed, in case it has changed you can just ignore the tick event and return.

IntegratedQuantum commented on src/blocks.zig:
```diff
@@ -371,35 +388,59 @@ pub const Block = packed struct { // MARK: Block
-				hashMap.putNoClobber(declaration.name, &@field(TouchFunctions, declaration.name)) catch unreachable;
-			}
+	pub fn loadFromZon(zon: ZonElement) ?TickEvent {
+		const _functionName = zon.get(?[]const u8, "name", null);
+		if(_functionName == null) return null;
```
```suggestion
		const _functionName = zon.get(?[]const u8, "name", null) orelse return null;
```

IntegratedQuantum commented on src/blocks.zig:
```diff
@@ -371,35 +388,59 @@ pub const Block = packed struct { // MARK: Block
+		const _functionName = zon.get(?[]const u8, "name", null);
+		if(_functionName == null) return null;
+
+		const _function = tickFunctions.getFunctionPointer(_functionName.?);
+		if(_function) |function| {
+			return TickEvent{.function = function, .chance = zon.get(f32, "chance", 1)};
 		}
```
```suggestion
		const function = tickFunctions.getFunctionPointer(_functionName.?) orelse {
			std.log.err("Could not find TickFunction {s}.", .{_functionName.?});
			return null;
		};
		return TickEvent{.function = function, .chance = zon.get(f32, "chance", 1)};
```

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -428,6 +432,9 @@ pub const ServerWorld = struct { // MARK: ServerWorld
 	doGameTimeCycle: bool = true,
 
+	tickSpeed: u32 = 12,
+	doTick: bool = true,
```
We don't really need this, it's redundant with a `tickSpeed` of 0.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1935,3 +1935,58 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+		}
+
+		pub fn getFunctionPointer(self: *Self, id: []const u8) ?*const Function {
+			std.debug.assert(id.len > 0);
```
You cannot assert this, this is a user provided value and might have length 0.
And this case is handled fine by the get function below anyways.

IntegratedQuantum replied:
Of course not. But the point is that we must never crash on bad user input.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1935,3 +1935,58 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+
+		pub fn getFunctionPointer(self: *Self, id: []const u8) ?*const Function {
+			std.debug.assert(id.len > 0);
+			const pointer = self.hashMap.getPtr(id) orelse {
+				return null;
+			};
+			return pointer.*;
```
```suggestion
			return self.hashMap.get(id);
```

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1935,3 +1935,58 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+	try std.testing.expect(testFunctions.hashMap.count() == 1);
+
+	const fnPtr = testFunctions.getFunctionPointer("testFunction");
+	try std.testing.expect(fnPtr != null);
```
Where possible, use `expectEqual` for better error messages on test failure.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1935,3 +1935,58 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+
+	const fnPtr = testFunctions.getFunctionPointer("testFunction");
+	try std.testing.expect(fnPtr != null);
+	if(fnPtr) |callback| {
+		try std.testing.expect(@TypeOf(callback) == *const TestFunction);
+	}
```
You already checked if it's null above.
Even without that you don't a null check when using `@TypeOf()`, which is evaluated at compile time.

And instead of comparing the types, you should compare the value with `&TestFunction.testFunction`, because that's what we are interested in here.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1935,3 +1935,58 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+	const TestFunction = fn(_: i32) void;
+	const TestFunctions = struct {
+		// Callback should register this
+		pub fn testFunction(_: i32) void {}
```
It would be nice to have a second function here, so checking if it's the correct function pointer has a failure case.

IntegratedQuantum reviewed (changes requested):
I checked the performance and it didn't even show up in the profiler, well done on that part!

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -1935,3 +1935,56 @@ fn CastFunctionReturnToAnyopaqueType(Fn: type) type {
+
+	const fnPtr = testFunctions.getFunctionPointer("testFunction");
+	try std.testing.expectEqual(&TestFunctions.testFunction, fnPtr.?);
+	try std.testing.expect(fnPtr.? != &TestFunctions.otherTestFunction);
```
That's testing that Zig uses different pointers for each function, which is not something we should worry about.

IntegratedQuantum reviewed (changes requested):
One more thing then we are ready to go here.

Also now would be great plan to think about future plans.
E.g. could you maybe make something like leaf decay as the first function?
It is important to add an actual in-game use case for system this quickly, so the code doesn't decay.

Furthermore one thing to consider in another PR would be a way to store data together with the function, so instead of having a `replaceWithCobblestone` function, we could have just a `replaceWith` function that stores the block type in another parameter.

IntegratedQuantum commented:
I also just tested this by making stone turn into cobblestone, with the provided function, and the performance is awful.
Not really a fault in your code (the main problem is the fact that it's remeshing, relighting and saving 200 chunks per tick).
However this is something that you should look into in the future. One possible solution would be to just select a subset of all chunks randomly, instead of updating all chunks in every tick. To balance this you could just check more blocks in the selected chunks.

IntegratedQuantum commented:
> Should this be done with some kind of persistant rotation mode?

Yeah, that's the best solution I think. Also persistent, not persistant.

Argmaster commented:
Will this be used for multi-state growing plants?

IntegratedQuantum commented:
Yeah, probably

ChibChi commented:
Tried the naive approach first. Breadth first search with a depth of around 4. Basically i check every branch untill I hit a block with the `cambium` tag. There is also the option to store the distance to a log in the leafs data, but I feel like this will get out of hand too with more complex trees.

IntegratedQuantum commented:
Maybe I wasn't clear enough: Please don't do it in this PR, as it's already quite full.
For this PR I was just waiting for you to fix the one small issue mentioned in https://github.com/PixelGuys/Cubyz/pull/1338#discussion_r2093365745

Argmaster commented:
I think leaf decay should be done when breaking logs. Leaves should three states: `wont-decay`, `will-decay` and `decaying`, player created ones should be `wont-decay`. When player breaks a block of wood you iterate a all the block in 5 radius sphere around that block and you set every `will-decay` leaves block to to `decaying` state. When you tick `will-decay` or `wont-decay` leaves block it is a noop, but for `decaying` leaves you check if there is log block in 5 block radius and if there is not you replace that leaves block with air. To make it fast, I think there should be multiple tiers of tick speeds for blocks, but with tick speeds relative to one gamerule or sth, basically x1 tier, x2 tier, x4 tier.

Argmaster commented:
Different approach - can we maybe support even scheduling with something like Command Design Pattern? Then what we could have an event that would check every block within radius X if its leaves and if it is has log block in range Y, if not it would instantly replace it, but we would do that as multiple events each one tick delayed, first check radius 1, then radius 2, then radius 3 etc. 
We would have to store events that were not yet executed but were scheduled with server, but for fixed set of programmed events it should be fine.

IntegratedQuantum commented:
> Different approach

Let's not make the first iteration more complicated than it needs to be, please.

Argmaster commented:
Oh no that was not my intention, just wanted to give suggestions, maybe its not the best place tho

IntegratedQuantum commented:
One very important question for leaves to keep in mind:
If none of the leaves you place manually can decay, then how are you supposed to build trees for the structure system?

Argmaster commented:
Magic I suppose. But seriously, you can use `/replace`

IntegratedQuantum commented:
> Magic I suppose. But seriously, you can use /replace

And if you forget? How I'm I supposed to spot such a mistake in review?
Years down the line we may get a bug report for leaves not despawning, and then what? Do we manually comb through all structure models?
